compiler/src/pipeline.rs

```diff
@@ -0,0 +1,489 @@
+use std::{
+    fmt::Display,
+    fs,
+    path::{Path, PathBuf},
+    time::Instant,
+};
+
+use analysis::convert_analyzed_to_pil_constraints;
+use ast::{
+    analyzed::Analyzed,
+    asm_analysis::AnalysisASMFile,
+    object::PILGraph,
+    parsed::{asm::ASMProgram, PILFile},
+    DiffMonitor,
+};
+use backend::{BackendType, Proof};
+use executor::{constant_evaluator, witgen::QueryCallback};
+use log::Level;
+use mktemp::Temp;
+use number::FieldElement;
+
+use crate::verify::{write_commits_to_fs, write_constants_to_fs, write_constraints_to_fs};
+
+pub struct GeneratedWitness<T: FieldElement> {
+    pub pil: Analyzed<T>,
+    pub constants: Vec<(String, Vec<T>)>,
+    pub witness: Option<Vec<(String, Vec<T>)>>,
+}
+
+pub struct PilWithConstants<T: FieldElement> {
+    pub pil: Analyzed<T>,
+    pub constants: Vec<(String, Vec<T>)>,
+}
+
+pub struct CompilationResult<T: FieldElement> {
+    /// Constant columns, potentially incomplete (if success is false)
+    pub constants: Vec<(String, Vec<T>)>,
+    /// Witness columns, potentially None (if success is false)
+    pub witness: Option<Vec<(String, Vec<T>)>>,
+    /// Proof, potentially None (if success is false)
+    pub proof: Option<Proof>,
+    /// Serialized low level constraints, potentially None (if success is false)
+    pub constraints_serialization: Option<String>,
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+enum Stage {
+    AsmFile,
+    AsmString,
+    Parsed,
+    Resolved,
+    AnalyzedAsm,
+    PilConstraints,
+    Graph,
+    Linked,
+    PilFile,
+    PilString,
+    AnalyzedPil,
+    OptimizedPil,
+    PilWithConstants,
+    GeneratedWitness,
+    Proof,
+}
+
+enum Artifact<T: FieldElement> {
+    AsmFile(PathBuf),
+    AsmString(Option<PathBuf>, String),
+    Parsed(Option<PathBuf>, ASMProgram<T>),
+    Resolved(ASMProgram<T>),
+    AnalyzedAsm(AnalysisASMFile<T>),
+    PilConstraints(AnalysisASMFile<T>),
+    Graph(PILGraph<T>),
+    Linked(PILFile<T>),
+    PilFile(PathBuf),
+    PilString(String),
+    AnaylyzedPil(Analyzed<T>),
+    OptimzedPil(Analyzed<T>),
+    PilWithConstants(PilWithConstants<T>),
+    GeneratedWitness(GeneratedWitness<T>),
+    Proof(CompilationResult<T>),
+}
+
+pub struct Pipeline<T: FieldElement> {
+    /// The current artifact. It is never None in practice, making it an Option is
+    /// only necessary so that we can take ownership of it in advance().
```

Comment:
so this is just a pointer? :D

Reply:
Not sure what you mean ^^ But basically, I would have expected I can do this in `Pipeline::advance`:

```rust
        self.artifact = match self.artifact {
            Artifact::AsmFile(path) => Artifact::AsmString(
                Some(path.clone()),
                fs::read_to_string(&path).unwrap(),
            ),
           // ...
```

But apparently, Rust won't let me and I need to do this:
```rust
        let artifact = std::mem::take(&mut self.artifact).unwrap();
        self.artifact = Some(match artifact {
        // ...
```